lib/ar_transaction_changes.rb

```diff
@@ -17,13 +18,25 @@ def transaction_changed_attributes
     @transaction_changed_attributes ||= HashWithIndifferentAccess.new
   end
 
-  def write_attribute(attr_name, value) # override
-    attr_name = attr_name.to_s
-    old_value = attributes[attr_name]
-    ret = super
-    unless transaction_changed_attributes.key?(attr_name) || value == old_value
-      transaction_changed_attributes[attr_name] = old_value
+  if ActiveRecord.gem_version >= Gem::Version.new("5.2.0.beta1")
+    def _write_attribute(attr_name, value) # override
+      attr_name = attr_name.to_s
+      old_value = attributes[attr_name]
+      ret = super
+      unless transaction_changed_attributes.key?(attr_name) || value == old_value
+        transaction_changed_attributes[attr_name] = old_value
+      end
+      ret
+    end
+  else
+    def write_attribute(attr_name, value) # override
```

Comment:
These two methods look the same.  Why not use the version check to determine the name then use define_method so the implementation doesn't need to be copied?  It requires explicitly passing arguments to super (e.g. `super(attr_name, value)`) but otherwise it doesn't seem like it will cause any problem.

Reply:
I was worried that using `define_method` would be significantly slower based on [this comment](https://github.com/rails/rails/blob/v5.2.0.beta2/activerecord/lib/active_record/attribute_methods/read.rb#L11-L12), but I should have measured it! Here's a benchmark:

```ruby
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"
  gem "rails", "~> 5.2.0.beta1"
  gem "ar_transaction_changes", path: "."
  gem "sqlite3"
  gem "benchmark-ips"
end

require "active_record"

ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end
end

class Post < ActiveRecord::Base
  include ArTransactionChanges
end

post = Post.new(id: 1)

Benchmark.ips do |x|
  x.report("attribute=") { post.id = post.id + 1 }
end
```

#### Baseline (without `ar_transaction_changes`)

```
Warming up --------------------------------------
          attribute=    31.248k i/100ms
Calculating -------------------------------------
          attribute=    347.426k (± 4.5%) i/s -      1.750M in   5.047349s
```

#### With `def`

```
Warming up --------------------------------------
          attribute=    17.451k i/100ms
Calculating -------------------------------------
          attribute=    188.077k (± 5.2%) i/s -    942.354k in   5.023675s
```

#### With `define_method`

```
Warming up --------------------------------------
          attribute=    16.717k i/100ms
Calculating -------------------------------------
          attribute=    183.604k (± 4.5%) i/s -    919.435k in   5.017719s
```

Using this library already makes `write_attribute` twice as slow (😬); any impact from `define_method` is negligible in comparison. I've pushed an updated commit that uses it.

Reply:
Thanks for benchmarking the library.  I hadn't done that before.

It looks like the lost performance is primarily from the `attributes[attr_name]` line which creates a new attributes hash each time it is called.  `self[attr_name]` would reduce the number of allocations, however, it still allocates a block for the `&block` in the `read_attribute` signature.  Using `_read_attribute` would avoid those object allocations and would be consistent with overriding `_write_attribute`.

Reply:
Nice find 👍 

I'm curious, why did you go with `read_attribute` over `_read_attribute` in https://github.com/dylanahsmith/ar_transaction_changes/commit/9c248f4a2988b8e7dd5ab6621f0f9b54a7e6bc3d?

Reply:
Mostly because it should work for newer and older rails versions.